File: python/src/fds/protobuf/stach/extensions/StachV3ExtensionFactory.py
import pyarrow as pa
import pyarrow.ipc as ipc
import base64
import pandas as pd
import json
from google.protobuf import json_format
from IPython.core.display import HTML
import io
from fds.protobuf.stach.extensions.v3.StachV3ArrowHelper import (
    _read_arrow_stream,
    _setup_level_columns,
    _parse_table_metadata as _parse_table_metadata_arrow,
    _parse_views_metadata as _parse_views_metadata_arrow,
    _parse_multi_level_headers as _parse_multi_level_headers_arrow,
    _build_column_headers_mapping as _build_column_headers_mapping_arrow,
    _populate_dataframe,
    _remove_ignored_columns,
    get_arrow_stream
)
from fds.protobuf.stach.extensions.v3.StachV3JsonHelper import (
    _parse_table,
    _parse_views,
    _parse_data_table,
    _parse_multi_level_headers,
    _build_column_headers_mapping,
    _build_data_table
)
from fds.protobuf.stach.extensions.v3.StachV3Constants import (
    group_prefix,
    level_name_column,
    security_name_column,
    column_separator,
    ignored_columns
)

# Module-level configuration variables (set before calling conversion functions)
column_headers_mapping = {}

def ConvertArrowFileToTable(file_path):
    global column_headers_mapping
    column_headers_mapping = {}  # Reset for each conversion
    data_set = {}

    # Read arrow file
    arrow_bytes, reader = _read_arrow_stream(file_path)
    schema = reader.schema

    # Setup level columns
    _setup_level_columns(arrow_bytes, column_headers_mapping)

    # Parse metadata
    view_headers = {}
    concat_columns_list = {}
    stach_table = None

    if schema.metadata is not None:
        metadata = {k.decode(): v.decode() for k, v in schema.metadata.items()}
        stach_table = _parse_table_metadata_arrow(metadata)
        view_headers = _parse_views_metadata_arrow(metadata)
        concat_columns_list = _parse_multi_level_headers_arrow(schema.metadata, stach_table, view_headers)

    # Build column headers mapping
    _build_column_headers_mapping_arrow(schema, view_headers, concat_columns_list, column_headers_mapping)

    # Populate DataFrame
    data_table = _populate_dataframe(reader, column_headers_mapping)

    # Remove ignored columns
    data_table = _remove_ignored_columns(data_table, ignored_columns)

    data_set['table'] = data_table

    return data_set

def ConvertArrowStreamToTable(arrow_bytes):
    global column_headers_mapping
    column_headers_mapping = {}  # Reset for each conversion
    data_set = {}

    # Get arrow stream from bytes
    reader = get_arrow_stream(arrow_bytes)
    if reader is None:
        raise Exception("Failed to create ArrowStreamReader from arrow_bytes")

    schema = reader.schema

    # Setup level columns
    _setup_level_columns(arrow_bytes, column_headers_mapping)

    # Parse metadata
    view_headers = {}
    concat_columns_list = {}
    stach_table = None

    if schema.metadata is not None:
        metadata = {k.decode(): v.decode() for k, v in schema.metadata.items()}
        stach_table = _parse_table_metadata_arrow(metadata)
        view_headers = _parse_views_metadata_arrow(metadata)
        concat_columns_list = _parse_multi_level_headers_arrow(schema.metadata, stach_table, view_headers)

    # Build column headers mapping
    _build_column_headers_mapping_arrow(schema, view_headers, concat_columns_list, column_headers_mapping)

    # Populate DataFrame
    data_table = _populate_dataframe(reader, column_headers_mapping)

    # Remove ignored columns
    data_table = _remove_ignored_columns(data_table, ignored_columns)

    data_set['table'] = data_table

    return data_set

def ConvertJsonToTable(jsonFileContent):
    global column_headers_mapping
    column_headers_mapping = {}  # Reset for each conversion
    data_set = {}

    if jsonFileContent is None:
        return
    stachJObject = json.loads(jsonFileContent)
    if stachJObject is None:
        return

    stachTable = _parse_table(stachJObject)
    customStachV3Views = _parse_views(stachJObject)
    customStachV3Table = _parse_data_table(stachJObject)
    multiLevelHeaders = _parse_multi_level_headers(stachJObject)
    columnHeadersMapping = _build_column_headers_mapping(stachTable, customStachV3Table, customStachV3Views, multiLevelHeaders)
    data_table = _build_data_table(customStachV3Table, columnHeadersMapping)

    data_set['table'] = data_table
    return data_set

File: dotnet/StachExtensions/FactSet.Protobuf.Stach.Extensions/StachV3ExtensionFactory.cs
using FactSet.Protobuf.Stach.Extensions.V3;
using Newtonsoft.Json.Linq;
using System.Collections.Generic;
using System.Data;
using System.IO;
using System.Linq;

namespace FactSet.Protobuf.Stach.Extensions
{
    public static class StachV3ExtensionFactory
    {  
        public static DataSet ConvertJsonToTable(string jsonFileContent)
        {
            var dataSet = new DataSet();

            if (string.IsNullOrEmpty(jsonFileContent))
            {
                return dataSet;
            }

            var stachJObject = JObject.Parse(jsonFileContent);
            if (stachJObject == null)
            {
                return null;
            }

            var dataTable = new DataTable();
            var columnHeadersMapping = new Dictionary<string, string>();

            // Parse schema components
            var stachTable = StachV3JsonHelper.ParseTableSchema(stachJObject);
            var customStachV3Views = StachV3JsonHelper.ParseViewsSchema(stachJObject);
            var customStachV3Table = StachV3JsonHelper.ParseColumnsAndRows(stachJObject);
            var multiLevelHeaders = StachV3JsonHelper.ParseMultiLevelHeaders(stachJObject);

            // Build column headers mapping
            StachV3JsonHelper.AddGroupColumns(customStachV3Table.Rows, columnHeadersMapping);
            var concatColumnsList = StachV3JsonHelper.BuildConcatenatedColumns(multiLevelHeaders, stachTable, customStachV3Views);
            StachV3JsonHelper.MapColumnHeaders(customStachV3Table, customStachV3Views, concatColumnsList, columnHeadersMapping);

            // Create DataTable structure
            StachV3JsonHelper.CreateDataTableColumns(dataTable, columnHeadersMapping);
            StachV3JsonHelper.PopulateDataTableRows(dataTable, customStachV3Table, columnHeadersMapping);
            StachV3JsonHelper.RemoveIgnoredColumns(dataTable);

            dataSet.Tables.Add(dataTable);
            return dataSet;
        }

        public static DataSet ConvertArrowStreamToTable(byte[] arrowBytes)
        {
            var ctx = new ConversionContext();

            // using ensures the reader (and its underlying MemoryStream) is disposed after conversion
            using (var reader = StachV3ArrowHelper.GetArrowStream(arrowBytes))
            {
                if (reader == null)
                    return null;

                var schema = reader.Schema;

                StachV3ArrowHelper.BuildGroupColumns(arrowBytes, ctx);

                if (schema?.HasMetadata == true)
                {
                    var metadata = (IDictionary<string, string>)schema.Metadata;
                    StachV3ArrowHelper.ParseTableMetadata(metadata, ctx);
                    StachV3ArrowHelper.ParseViewsMetadata(metadata, ctx);
                    StachV3ArrowHelper.BuildConcatColumnsList(metadata, ctx);
                    StachV3ArrowHelper.MapColumnHeaders(schema.FieldsList.ToList(), ctx);
                }

                StachV3ArrowHelper.BuildDataTableColumns(ctx);
                StachV3ArrowHelper.PopulateDataTable(reader, ctx);
                StachV3ArrowHelper.RemoveIgnoredColumns(ctx);
            }

            var dataSet = new DataSet();
            dataSet.Tables.Add(ctx.DataTable);
            return dataSet;
        }
        
        public static DataSet ConvertArrowFileToTable(string filePath)
        {
            if (!File.Exists(filePath))
                return null;

            byte[] arrowBytes = File.ReadAllBytes(filePath);
            return ConvertArrowStreamToTable(arrowBytes);
        }
    }
}